### ISSUE TITLE
Data for testing against interfaces

### ISSUE THREAD

**Issue by Max-Goebel** [oasp/oasp4j#212]
How will we treat test data for testing the server against its interfaces to client and to external systems? 
Typically, these test data must be available as XML at test runtime. The first question is, when and how this XML should be created:
* Solution 1: Manually, when the test is implemented.
* Solution 2: Automated with data generator, when the test is implemented. 
* Solution 3: Automated with data generator, when the test is executed (on the fly). Hint: As external systems are replaced by doubles (see https://github.com/oasp/oasp4j/wiki/guide-testing#integration-levels ), the generator may have to be integrated into the double.

If we want to use a data generator, there are different variants:
* Generator gets an XML template and a configuration file. It replaces some fields in the XML templates with values from the configuration file.
* Generator gets a configuration file only and generates the complete XML file.
* Maybe we can use an existing generation tool.

The purpose of this ticket is to clear these questions:
* Which solution is the best?
* How should this solution be supported by OASP?

**Comment by hohwille** [comment]
why do you say that such testdata must be XML?
The most important aspect in that case is typically the initial content of the database. How we maintain and import testdata into database is yet to be defined. There were people suggesting dumps (binary or SQL) as well as snapshots. With DBUnit we can also use XML. Besides the format to import is also the format to maintain what must not be the same.

**Comment by Max-Goebel** [comment]
For discussing database-testdata we should create a seperate ticket or use https://github.com/oasp/oasp4j/issues/153. For discussing testdata used for testing Java-methods there is also a ticket: https://github.com/oasp/oasp4j/issues/207. Of course there are correlations between interface-testdata, db-testdata and java-testdata. But I think we better should have three tickets instead of one monster ticket.

**Comment by may-bee** [comment]
Are interface-testdata for you the testdata with will be part of the service call for example?

**Comment by Max-Goebel** [comment]
Yes. I mean testdata for [subsystem-test](https://github.com/oasp/oasp4j/wiki/guide-testing#level-3-subsystem-test).

**Comment by may-bee** [comment]
In my previous projects we again stick to the builder pattern in this case. We have worked with REST + XML and thus we build up test data using builder implementations for the JAXB Objects we have to transmit to the service. Afterwards we marshall the objects to XML for transmission.
Same pattern - refactoring stable test data creation.

**Comment by Max-Goebel** [comment]
To be more specific, there are the following types of test data that are needed for testing interfaces:

Input data:
* XML, used for sending SOAP requests to the system under test
* URIs, used for sending REST requests to the system under test
* Serialized method calls, used for sending HTTP-invoker requests to the system under test

Expected results: 
* XML, used for validating SOAP requests to service providers sent by the system under test 
* JSON, used for validating REST-responses to a JS-client sent by the system under test
* Serialized method calls, used for validating HTTP-invoker requests to backend-systems sent by the system under test
* Serialized method call responses, used for validating HTTP-invoker responses to a backend-system sent by the system under test

Of course, there are more protocols and data formats that are relevant in the real world. But at this point in time, we should confine ourselves on the protocols and formats described in the OASP4J architecture guide (https://github.com/oasp/oasp4j/wiki/guide-service-layer#types-of-services ).

@may-bee: Do I guess right, that you propose solution 3 (on-the-fly-generation) using the builder pattern for all these types of input data and expected results?
@All others: What do you think?

**Comment by may-bee** [comment]
I am not quite sure whether I am understanding the difference between solution 2 & 3.
My viewpoint on the difference is, that solution 2 depends on declared test data sets specified during implementation of the test, whereas solution 3 might generate test data randomly while execution.

If I understood this correctly, my comment mainly focuses on Builder-Pattern with solution 2 as you strictly specify the test data to be used during execution. Nevertheless, there might be also properties, which will be filled with random values.

**Comment by Max-Goebel** [comment]
Solution 2 means, that you create XML (or JSON or any other interface format) files during test implementation and keep these files under version control. With solution 3 you implement Java code that creates the XML (or other interface format) data at runtime, e.g. with JAXB. 

If you want to generate data randomly is a different question. With both solutions you could create random data, but I do not think that this is desirable (except for performance testing maybe).
